opm/autodiff/BlackoilModelBase_impl.hpp

```diff
@@ -1694,6 +1695,108 @@ namespace detail {
+                //or a THP control that specifies what we need.
+                //Pick the value that gives the most restrictive flow
+                for (int ctrl_index=0; ctrl_index < nwc; ++ctrl_index) {
+
+                    if (well_controls_iget_type(ctrl, ctrl_index) == BHP) {
+                        bhps[w] = well_controls_iget_target(ctrl, ctrl_index);
+                    }
+
+                    if (well_controls_iget_type(ctrl, ctrl_index) == THP) {
+                        double aqua = 0.0;
+                        double liquid = 0.0;
+                        double vapour = 0.0;
+
+                        if (active_[ Water ]) {
+                            aqua = well_state.wellRates()[w*np + pu.phase_pos[ Water ] ];
+                        }
+                        if (active_[ Oil ]) {
+                            liquid = well_state.wellRates()[w*np + pu.phase_pos[ Oil ] ];
+                        }
+                        if (active_[ Gas ]) {
+                            vapour = well_state.wellRates()[w*np + pu.phase_pos[ Gas ] ];
+                        }
+
+                        const int vfp        = well_controls_iget_vfp(ctrl, ctrl_index);
+                        const double& thp    = well_controls_iget_target(ctrl, ctrl_index);
+                        const double& alq    = well_controls_iget_alq(ctrl, ctrl_index);
+
+                        //Set *BHP* target by calculating bhp from THP
+                        const WellType& well_type = wells().type[w];
+
+                        const double gravity = detail::getGravity(geo_.gravity(), UgGridHelpers::dimensions(grid_));
+
+                        if (well_type == INJECTOR) {
+                            double dp = wellhelpers::computeHydrostaticCorrection(
+                                        wells(), w, vfp_properties_.getInj()->getTable(vfp)->getDatumDepth(),
+                                        stdWells().wellPerforationDensities(), gravity);
+                            const double bhp = vfp_properties_.getInj()->bhp(vfp, aqua, liquid, vapour, thp) - dp;
+                            // apply the strictes of the bhp controlls i.e. smallest bhp for injectors
+                            if ( bhp < bhps[w]) {
+                                bhps[w] = bhp;
+                            }
+                        }
+                        else if (well_type == PRODUCER) {
+                            double dp = wellhelpers::computeHydrostaticCorrection(
+                                        wells(), w, vfp_properties_.getProd()->getTable(vfp)->getDatumDepth(),
+                                        stdWells().wellPerforationDensities(), gravity);
+
+                            const double bhp = vfp_properties_.getProd()->bhp(vfp, aqua, liquid, vapour, thp, alq) - dp;
+                            // apply the strictes of the bhp controlls i.e. largest bhp for injectors
```

Comment:
Here "injectors" should be "producers". Also strictes -> strictest, controlls -> controls here and above.